src/java.base/share/classes/java/lang/Math.java

```diff
@@ -1187,6 +1188,29 @@ public static long multiplyHigh(long x, long y) {
         }
     }
 
+    /**
+     * Returns as a {@code long} the most significant 64 bits of the unsigned
+     * 128-bit product of two unsigned 64-bit factors.
+     *
+     * @param x the first value
+     * @param y the second value
+     * @return the result
+     * @see #multiplyHigh
+     * @since 18
+     */
+    public static long unsignedMultiplyHigh(long x, long y) {
+        long x0 = x & 0xffffffffL;
+        long x1 = x >>> 32;
+        long y0 = y & 0xffffffffL;
+        long y1 = y >>> 32;
+
+        long t = x1 * y0 + ((x0 * y0) >>> 32);
+        long z0 = x0 * y1 + (t & 0xffffffffL);
+        long z1 = t >>> 32;
+
+        return x1 * y1 + z1 + (z0 >>> 32);
```

Comment:
```suggestion
        long result = Math.multiplyHigh(x, y);
        if (x < 0) result += y;
        if (y < 0) result += x;
        return result;
```

Reply:
This is just subtracting the 2's-complement offset. I guess the idea, longer term, is that this be an intrinsic anyway, but if you do `unsignedMultiplyHigh` this way you'll utilize the existing `multiplyHigh` intrinsic on all platforms that have it.

Reply:
You can also do that branchlessly which might prove better

         long result = Math.multiplyHigh(x, y);
         result += (y & (x >> 63));
         result += (x & (y >> 63));
         return result;

Reply:
> You can also do that branchlessly which might prove better
> 
> ```
>      long result = Math.multiplyHigh(x, y);
>      result += (y & (x >> 63));
>      result += (x & (y >> 63));
>      return result;
> ```
I doubt very much that it would be better, because these days branch prediction is excellent, and we also have conditional select instructions. Exposing the condition helps C2 to eliminate it if the range of args is known. The `if` code is easier to understand.

Benchmark results, with one of the operands changing signs every iteration, 1000 iterations:

```
Benchmark                  Mode  Cnt     Score    Error  Units
MulHiTest.mulHiTest1       avgt    3  1570.587 ± 16.602  ns/op
MulHiTest.mulHiTest2       avgt    3  2237.637 ±  4.740  ns/op
```
In any case, note that with this optimization the unsigned mulHi is in the nanosecond range, so Good Enough. IMO.

Reply:
But weirdly, it's the other way around on AArch64, but there's little in it:

```
Benchmark             Mode  Cnt     Score   Error  Units
MulHiTest.mulHiTest1  avgt    3  1492.108 ± 0.301  ns/op
MulHiTest.mulHiTest2  avgt    3  1219.521 ± 1.516  ns/op
```

but this is only in the case where we have unpredictable branches. Go with simple and easy to understand; it doesn't much matter.